src/hotspot/share/gc/shenandoah/heuristics/shenandoahGlobalHeuristics.cpp

```diff
@@ -101,31 +125,50 @@ void ShenandoahGlobalHeuristics::choose_global_collection_set(ShenandoahCollecti
   for (size_t idx = 0; idx < size; idx++) {
     ShenandoahHeapRegion* r = data[idx]._region;
     if (cset->is_preselected(r->index())) {
+      fatal("There should be no preselected regions during GLOBAL GC");
       continue;
     }
     bool add_region = false;
-    if (r->is_old()) {
+    if (r->is_old() || (r->age() >= tenuring_threshold)) {
       size_t new_cset = old_cur_cset + r->get_live_data_bytes();
+      if ((r->garbage() > garbage_threshold)) {
+        while ((new_cset > max_old_cset) && (unaffiliated_young_regions > 0)) {
+          unaffiliated_young_regions--;
+          regions_transferred_to_old++;
+          max_old_cset += region_size_bytes / ShenandoahOldEvacWaste;
+        }
+      }
       if ((new_cset <= max_old_cset) && (r->garbage() > garbage_threshold)) {
         add_region = true;
         old_cur_cset = new_cset;
       }
-    } else if (r->age() < tenuring_threshold) {
+    } else {
+      assert(r->is_young() && (r->age() < tenuring_threshold), "DeMorgan\'s law");
```

Comment:
Not quite DeMorgan's law because `r` could be unaffiliated (it isn't _supposed_ to be, of course).

Reply:
I'm fixing the message here.